gen/mcat2config.ml

```diff
@@ -0,0 +1,729 @@
+            raise (Failed (sprintf "Unknown Variable: %s\n" name))
+        | (Lrs _ as h) :: h2 :: t ->
+          let a1, a2, _, _ = extract_diratom prev h2 in
+            [
+              {
+                node_val = E.{ edge = Po (Same, Dir W, Dir R); a1; a2 };
+                branch_list = f (h2 :: t) h;
+              };
+            ]
+        | h::h2::t ->
+          let branch_list a2 =
+            if a2 = None then f (h2 :: t) h
+            else f t h
+          in
+          match_edge h prev h2 branch_list
+        | h::[] -> match_edge h prev empty_expr (fun _ -> [])
+        | [] -> raise (Failed "cannot provide empty list")
+      in f expl empty_expr
+    
+    and listofnode nodel = 
+      (*translate edge_node list into edge list list for printing
+        *)
+      let rec f node =
+        match node.branch_list with
+        | h::t ->
+          List.concat (List.map (fun a -> 
+                       List.map (fun b -> node.node_val::b)
+                       (f a)
+                       ) (h::t))
+        | [] -> [[node.node_val]] in
+      List.concat (List.map f nodel)
+    
+    and check_prev a = 
+      match a with
+      | Empty _ -> None
+      | An (b,_) -> Some b
+      | Dd (_,an,_) -> an
+      | _ -> None
+    and check_if_dir a =
+      match a with
+      | D (a,_) ->  Dir a
+      | Ext (NoDir,_) -> NoDir
+      | Dd (r,_,_) -> r
+      | _ -> Irr
+    
+    and match_inter expl =
+      match expl with
+      | h::h2::h3::[] -> begin
+        match (check_if_dir h),h2,h3 with
+        | Dir R, PTE _, Imp _ -> Dd (Dir R, Some (A.Pte A.Read, None), "ImpPTE Read")
+        | Dir R, An ((A.Tag, None),_), Imp _->Dd (Dir R, Some (A.Tag, None), "ImpTag Read")
+        | _ -> raise (Failed "wrong inter")
+      end
+      | _ -> raise (Failed "wrong inter")
+   
+    and unroll_inter expr =
+    let open AST in
+      let f l =
+        match l with 
+        | Op2 (Inter, expl) -> expl
+        | _ -> [l] in
+      match expr with
+      | Op2 (Inter, exp) -> Op2 (Inter,(List.concat (List.map f exp)))
+      | _ -> expr
+    
+    and extract_diratom prev next =
+      let extr1 = check_if_dir prev in
+      let extr2 = check_if_dir next in
+      
+      let a1 = check_prev prev in
+      let a2 = check_prev next in
+      let extr1 = check_dirfroma a1 extr1 in
+      let extr2 = check_dirfroma a2 extr2 in
+      a1,a2,extr1,extr2
+        
+    (*
+    ------------------------------------------
+    entry point for constructing tree
+    ------------------------------------------   
+    *)
+    and get_tree ast = 
+      let rec tree f l = 
+        match l with
+        | h::t -> begin try
+          (f h)::(tree f t)
+        with
+          | Failed s -> begin 
+            if O.verbose > 0 then 
+              (eprintf "%s" s)
+          end;
+            (tree f t)
+          | _ -> (tree f t)
+        end
+        | [] -> [] in
+      let tree_base = tree get_ins ast in
```

Comment:
that's correct, I hadn't yet added fr and co to my matching function, this should now be done. Thanks for noticing! I have also attempted to implement some lazy evaluation like you suggested, this can be seen in the map_vars function on line 539. Please let me know if this is not what you intended :)